File: ibkr_show_backend/app/api/routes/cash_flows.py
from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.api.deps import get_cash_flow_service, require_authenticated_session
from app.core.auth import AuthSession
from app.clients.es_client import ESClientError
from app.schemas.cash_flows import CashFlowListResponse, CashFlowSummaryResponse
from app.services.cash_flow_service import CashFlowService

router = APIRouter(prefix="/cash-flows", tags=["cash-flows"])


@router.get("", response_model=CashFlowListResponse)
def list_cash_flows(
    start_date: str | None = Query(default=None),
    end_date: str | None = Query(default=None),
    currency: str | None = Query(default=None),
    flow_direction: str | None = Query(default=None),
    sort_by: str = Query(default="date_time"),
    sort_order: str = Query(default="desc"),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=200),
    _auth_session: AuthSession = Depends(require_authenticated_session),
    service: CashFlowService = Depends(get_cash_flow_service),
) -> CashFlowListResponse:
    try:
        return service.list_cash_flows(
            start_date=start_date,
            end_date=end_date,
            currency=currency,
            flow_direction=flow_direction,
            sort_by=sort_by,
            sort_order=sort_order,
            page=page,
            page_size=page_size,
        )
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    except ESClientError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc


@router.get("/summary", response_model=CashFlowSummaryResponse)
def get_cash_flow_summary(
    start_date: str | None = Query(default=None),
    end_date: str | None = Query(default=None),
    currency: str | None = Query(default=None),
    flow_direction: str | None = Query(default=None),
    _auth_session: AuthSession = Depends(require_authenticated_session),
    service: CashFlowService = Depends(get_cash_flow_service),
) -> CashFlowSummaryResponse:
    try:
        return service.summarize_cash_flows(
            start_date=start_date,
            end_date=end_date,
            currency=currency,
            flow_direction=flow_direction,
        )
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    except ESClientError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc

File: ibkr_show_frontend/src/api/cashFlows.ts
import { request } from './http'
import type { CashFlowListResponse, CashFlowSummaryResponse } from '@/types/cashFlows'

export interface CashFlowQuery {
  start_date?: string
  end_date?: string
  currency?: string
  flow_direction?: string
  sort_by?: string
  sort_order?: 'asc' | 'desc'
  page?: number
  page_size?: number
}

export function fetchCashFlows(params: CashFlowQuery): Promise<CashFlowListResponse> {
  const searchParams = new URLSearchParams()

  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== '') {
      searchParams.set(key, String(value))
    }
  })

  const queryString = searchParams.toString()
  return request<CashFlowListResponse>(`/api/cash-flows${queryString ? `?${queryString}` : ''}`)
}

export function fetchCashFlowSummary(
  params: Omit<CashFlowQuery, 'sort_by' | 'sort_order' | 'page' | 'page_size'>,
): Promise<CashFlowSummaryResponse> {
  const searchParams = new URLSearchParams()

  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== '') {
      searchParams.set(key, String(value))
    }
  })

  const queryString = searchParams.toString()
  return request<CashFlowSummaryResponse>(`/api/cash-flows/summary${queryString ? `?${queryString}` : ''}`)
}
